Sources/AsyncAlgorithms/AsyncWithLatestFromSequence.swift

```diff
@@ -0,0 +1,184 @@
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+public extension AsyncSequence {
+  /// Combines `self` with another ``AsyncSequence`` into a single ``AsyncSequence`` where each
+  /// element from `self` is aggregated to the latest known element from the `other` sequence (if any) as a tuple.
+  ///
+  /// Remark: as the `other` sequence is being iterated over in the context of its own ``Task``, there is no guarantee
+  /// that its latest know element is the one that has just been produced when the base sequence produces its next element.
+  ///
+  /// ```
+  /// let base = AsyncChannel<Int>()
+  /// let other = AsyncChannel<String>()
+  /// let sequence = base.withLatest(from: other)
+  ///
+  /// Task {
+  ///   for element in await sequence {
+  ///    print(element)
+  ///   }
+  /// }
+  ///
+  /// await other.send("a")
+  /// await other.send("b")
+  ///
+  /// ... later in the application flow
+  ///
+  /// await base.send(1)
+  ///
+  /// // will print: (1, "b")
+  /// ```
+  ///
+  /// - Parameter other: the other ``AsyncSequence``
+  /// - Returns: an ``AsyncWithLatestFromSequence`` where elements are a tuple of an element from `self` and the
+  /// latest known element (if any) from the `other` sequence.
+  func withLatest<Other: AsyncSequence>(
+    from other: Other
+  ) -> AsyncWithLatestFromSequence<Self, Other> {
+    AsyncWithLatestFromSequence(self, other)
+  }
+}
+
+/// ``AsyncWithLatestFromSequence`` is an ``AsyncSequence`` where elements are a tuple of an element from `base` and the
+/// latest known element (if any) from the `other` sequence.
+public struct AsyncWithLatestFromSequence<Base: AsyncSequence, Other: AsyncSequence>: AsyncSequence {
+  public typealias Element = (Base.Element, Other.Element)
+  public typealias AsyncIterator = Iterator
+
+  let base: Base
+  let other: Other
+
+  init(_ base: Base, _ other: Other) {
+    self.base = base
+    self.other = other
+  }
+
+  public func makeAsyncIterator() -> Iterator {
+    Iterator(
+      base: self.base.makeAsyncIterator(),
+      other: self.other
+    )
+  }
+
+  public struct Iterator: AsyncIteratorProtocol {
+    enum OtherState {
+      case notStarted
+      case idle
+      case element(Result<Other.Element, Error>)
+    }
+
+    enum BaseDecision {
+      case pass
+      case returnElement(Result<Element, Error>)
+    }
+
+    var base: Base.AsyncIterator
+    let other: Other
+    let otherState: ManagedCriticalState<OtherState>
+    var otherTask: Task<Void, Never>?
+    var isTerminated: ManagedCriticalState<Bool>
+
+    // for testability purpose
+    var onBaseElement: (@Sendable (Base.Element) -> Void)?
+    var onOtherElement: (@Sendable (Other.Element?) -> Void)?
```

Comment:
I strongly believe we shouldn't insert code into these things just for testing. The behaviour should be fully observable from the public APIs.

Reply:
@phausler do you have an opinion about that ?

Reply:
the more exposed APIs we have the more potential bugs/bad-interactions there could be. If we need some sort of mechanism for testing then we should figure out a way to observe the issue via the tests. Since the tests can control the stepping of the global executors, and we also control both the inputs AND the demand, we should be able to observe everything that is of importance in the tests.